### PR TITLE
Implemented a clone of the Microprofile OpenAPI spec and implementation but for AsyncAPI.

### PR DESCRIPTION
**Pull request by EricWittmann**
Signed-off-by: Eric Wittmann <eric.wittmann@gmail.com>

### CODE REVIEW

**Comment by kenfinnigan** [comment]
Silly question, would there be the possibility of having the same "engine" for OpenAPI and AsyncAPI with just different surrounding processing?

**Comment by kenfinnigan** [comment]
FYI @cescoffier

**Comment by EricWittmann** [comment]
Yes, I think there is a lot to share between the two if we think this is a good idea.  I didn't make much of an attempt at doing that, but instead tried to get something useful going asap.

**Comment by kenfinnigan** [comment]
Totally fine, was more a crazy thought. We can always look to refactor to a common core in the future

**Comment by EricWittmann** [comment]
The structure of the modules is the same as for smallrye-open-api.  The difference is that there are two extra modules which represent the bits that would normally be in eclipse rather than here (the spec and the spec tck).

**Comment by EricWittmann** [comment]
Also note that all of the data models and I/O is coming from here:  https://github.com/apicurio/apicurio-data-models

That's what I was waiting on before bootstrapping this project.  I didn't want to implement the data model in both places like I did for OpenAPI.  Thanks to @jsenko for implementing most of the AsyncAPI data model in `apicurio-data-models`.

Any official release of this stuff into MP would require that we divorce the `apicurio-data-models` stuff (while still re-using it somehow).  That's likely an annoying lift involving introducing lots of data model interfaces and factories.

**Comment by kenfinnigan** [comment]
Does it make sense for the data models to be in SmallRye, used by Apicurio and other SmallRye projects?

**Comment by EricWittmann** [comment]
Possibly.  But the spec would need some sort of notion of the data models (the MP OpenAPI spec has a bunch of interfaces for the models).  That's the annoying part IMO.

**Comment by EricWittmann** [comment]
I'm referring to these when I talk about data model interfaces:  https://github.com/eclipse/microprofile-open-api/tree/master/api/src/main/java/org/eclipse/microprofile/openapi/models

**Comment by kenfinnigan** [comment]
oh I see, so the MP spec has interfaces.

What about if we split SmallRye OpenAPI into pieces. There was a module that just implemented the model interfaces, which could then be used by everything?

**Comment by kenfinnigan** [comment]
Possibly as a precursor to getting those models moved to a specific MP spec at some point?

**Comment by EricWittmann** [comment]
Unfortunately it's slightly more complicated than that due to some Apicurio requirements that aren't readily apparent (and is why `apicurio-data-models` exists as a project).  But yeah, there is certainly a sensible way to organize all of this for maximum re-use.  We just need to hash it out and then do the work.  But I didn't want to tackle any of that until some messaging experts had a chance to review this and decide whether it's useful at all. :)

File: implementation/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/target/
```
Comment (kenfinnigan) [inline implementation/.gitignore]: Shouldn't need sub directory .gitignore files

File: implementation/pom.xml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>javax.json.bind</groupId>
+            <artifactId>javax.json.bind-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.jboss.resteasy</groupId>
+            <artifactId>resteasy-core-spi</artifactId>
```
Comment (kenfinnigan) [inline implementation/pom.xml]: Why the RESTEasy core spi in test scope?

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,222 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!-- ~ Copyright 2018 Red Hat, Inc. ~ ~ Licensed under the Apache License, Version 2.0 (the "License"); ~ you may
```
Comment (kenfinnigan) [inline pom.xml]: Can remove copyright from `pom.xml` files

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.smallrye</groupId>
+        <artifactId>smallrye-parent</artifactId>
+        <version>7</version>
```
Comment (kenfinnigan) [inline pom.xml]: `8` I think is the latest now

File: spec-tck/pom.xml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        <groupId>io.smallrye</groupId>
+        <artifactId>smallrye-async-api-parent-1.0</artifactId>
+        <version>1.0.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>smallrye-async-api-spec-tck-1.0</artifactId>
+    <name>SmallRye: MicroProfile AsyncAPI Spec TCK</name>
+
+    <properties>
+        <arquillian.version>1.1.13.Final</arquillian.version>
+        <restclient.version>1.1</restclient.version>
```
Comment (kenfinnigan) [inline spec-tck/pom.xml]: Should this be `1.3` now that MP 3.0 is out?

File: spec/pom.xml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.smallrye</groupId>
+        <artifactId>smallrye-async-api-parent-1.0</artifactId>
+        <version>1.0.0-SNAPSHOT</version>
+    </parent>
+
+    <artifactId>smallrye-async-api-spec-1.0</artifactId>
```
Comment (kenfinnigan) [inline spec/pom.xml]: Maybe switch it to be `smallrye-async-api-spec-api`?

File: implementation/.gitignore (hunk 1 above)
Comment (EricWittmann) [reply implementation/.gitignore]: Agreed.  Will change it.

File: implementation/pom.xml (hunk 2 above)
Comment (EricWittmann) [reply implementation/pom.xml]: Probably copy/paste leftover from open-api.  Can't recall why it's needed there though - would need to look into that.

File: pom.xml (hunk 3 above)
Comment (EricWittmann) [reply pom.xml]: OK will do.

**Comment by EricWittmann** [comment]
@kenfinnigan I think I've addressed all/most of the changes you requested.  Have a look - happy to make other changes if required.

**Pull request merged**